conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/DynamicInterpreter.scala

```diff
@@ -214,43 +213,20 @@ object DynamicInterpreter {
     * @return The list of lists of conformance rule groups
     */
   private def groupMappingRules(rules: List[ConformanceRule], schema: StructType): List[List[ConformanceRule]] = {
-    val groups = new ListBuffer[List[ConformanceRule]]
-    var group = new ListBuffer[ConformanceRule]
-    var lastArrayParent = ""
-
-    def pushGroup(arrayPath: String = ""): Unit = {
-      if (group.nonEmpty) {
-        groups += group.toList
-        group = new ListBuffer[ConformanceRule]
-      }
-      lastArrayParent = arrayPath
-    }
-
-    for (rule <- rules) {
-      rule match {
-        case m: MappingConformanceRule =>
-          val arrayParent = SchemaUtils.getDeepestArrayPath(schema, rule.outputColumn)
-          arrayParent match {
-            case Some(arr) =>
-              if (lastArrayParent != arr) {
-                pushGroup(arr)
-              }
-              group += m
-            case None =>
-              pushGroup()
-              groups += List(m)
-          }
-
-          // Something
-        case a =>
-          pushGroup()
-          groups += List(a)
-      }
-    }
-    pushGroup()
-    groups.toList
+    Algorithms.stableGroupByOption[ConformanceRule, String](rules, {
+      case m: MappingConformanceRule => SchemaUtils.getDeepestArrayPath(schema, m.outputColumn)
+      case _ => None
+    }).map(_.toList).toList
```

Comment:
I get that you are mapping `Seq` to `List` here, but why not have the more abstract return type in the function: `Seq[Seq[ConformanceRule]]` we are not making use of some `List`-specific properties with the return of this function (we just `foldLeft` from what I could trace). I'd substitute it for the more abstract `Seq` type.

Reply:
Also, the implementation underneath is already a `List` when you look at the implementation of this algorithm.